Allow setting index.creation_date on index creation for plugin compatibility and migrations

AndreKurait opened the pull request:
<!--  Thanks for sending a pull request, here are some tips:

1. If this is a fix for an undisclosed security vulnerability, please STOP. All security vulnerability reporting and fixes should be done as per our security policy https://github.com/opensearch-project/OpenSearch/security/policy
2. If this is your first time, please read our contributor guidelines: https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md and developer guide https://github.com/opensearch-project/OpenSearch/blob/main/DEVELOPER_GUIDE.md
3. Ensure you have added or ran the appropriate tests for your PR: https://github.com/opensearch-project/OpenSearch/blob/main/TESTING.md
-->

### Description
Allow setting index.creation_date on index creation for plugin compatibility and migrations.

Aligns behavior with snapshot restore to let migrations retain original creation timestamps.

### Manual Testing

Verified with script which ran with output
```
➜  ~ ./validation.sh
=== CONFIG ===
Endpoint: http://localhost:9200
Suffix:   1762548057

=== 1) Create index WITHOUT explicit creation_date (test-index-auto-1762548057) ===
{
  "acknowledged": true,
  "shards_acknowledged": true,
  "index": "test-index-auto-1762548057"
}

--- GET settings (test-index-auto-1762548057) ---
{
  "test-index-auto-1762548057": {
    "settings": {
      "index": {
        "replication": {
          "type": "DOCUMENT"
        },
        "number_of_shards": "1",
        "provided_name": "test-index-auto-1762548057",
        "creation_date": "1762548057850",
        "number_of_replicas": "1",
        "uuid": "p9myRyg8S4SpIey0vMXIrw",
        "version": {
          "created": "137257827"
        }
      }
    }
  }
}
Requested: (auto-generated)
Actual:    1762548057850

=== 2) Create index WITH explicit creation_date (test-index-explicit-1762548057) ===
{
  "acknowledged": true,
  "shards_acknowledged": true,
  "index": "test-index-explicit-1762548057"
}

--- GET settings (test-index-explicit-1762548057) ---
{
  "test-index-explicit-1762548057": {
    "settings": {
      "index": {
        "replication": {
          "type": "DOCUMENT"
        },
        "number_of_shards": "1",
        "provided_name": "test-index-explicit-1762548057",
        "creation_date": "1234567890000",
        "number_of_replicas": "1",
        "uuid": "I3-vNRqLQz2xW4PKX3-Izw",
        "version": {
          "created": "137257827"
        }
      }
    }
  }
}
Requested: 1234567890000
Actual:    1234567890000

=== 3) Create INDEX TEMPLATE WITHOUT creation_date (no-date-template-1762548057) ===
{
  "acknowledged": true
}

--- GET template (no-date-template-1762548057) ---
{
  "index_templates": [
    {
      "name": "no-date-template-1762548057",
      "index_template": {
        "index_patterns": [
          "no-date-*1762548057"
        ],
        "template": {
          "settings": {
            "index": {
              "number_of_shards": "1"
            }
          }
        },
        "composed_of": []
      }
    }
  ]
}

--- Create index matching that template (no-date-1762548057) ---
{
  "acknowledged": true,
  "shards_acknowledged": true,
  "index": "no-date-1762548057"
}

--- GET settings (no-date-1762548057) ---
{
  "no-date-1762548057": {
    "settings": {
      "index": {
        "replication": {
          "type": "DOCUMENT"
        },
        "number_of_shards": "1",
        "provided_name": "no-date-1762548057",
        "creation_date": "1762548058741",
        "number_of_replicas": "1",
        "uuid": "LEAlsbeyQfSsCkAaV5sL7Q",
        "version": {
          "created": "137257827"
        }
      }
    }
  }
}
Requested: (auto-generated)
Actual:    1762548058741

=== 4) Create INDEX TEMPLATE WITH creation_date (with-date-template-1762548057) ===
{
  "acknowledged": true
}

--- GET template (with-date-template-1762548057) ---
{
  "index_templates": [
    {
      "name": "with-date-template-1762548057",
      "index_template": {
        "index_patterns": [
          "with-date-*1762548057"
        ],
        "template": {
          "settings": {
            "index": {
              "number_of_shards": "1",
              "creation_date": "1234567890000"
            }
          }
        },
        "composed_of": []
      }
    }
  ]
}

--- Create index matching that template (with-date-1762548057) ---
{
  "acknowledged": true,
  "shards_acknowledged": true,
  "index": "with-date-1762548057"
}

--- GET settings (with-date-1762548057) ---
{
  "with-date-1762548057": {
    "settings": {
      "index": {
        "replication": {
          "type": "DOCUMENT"
        },
        "number_of_shards": "1",
        "provided_name": "with-date-1762548057",
        "creation_date": "1234567890000",
        "number_of_replicas": "1",
        "uuid": "G_-aknqPSKaucpMWlK4QSw",
        "version": {
          "created": "137257827"
        }
      }
    }
  }
}
Requested: 1234567890000
Actual:    1234567890000

=== SUMMARY ===
Index                        | Source                 | Requested (ms)      | Actual (ms)         | Match
--------------------------------------------------------------------------------------------------------------
test-index-auto-1762548057   | direct (no date)       | (auto-generated)    | 1762548057850       | n/a  
test-index-explicit-1762548057 | direct (explicit date) | 1234567890000       | 1234567890000       | yes  
no-date-1762548057           | template (no date)     | (auto-generated)    | 1762548058741       | n/a  
with-date-1762548057         | template (explicit date) | 1234567890000       | 1234567890000       | yes  

Done.
```

### Check List
- [x] Functionality includes testing.
- [ ] API changes companion pull request [created](https://github.com/opensearch-project/opensearch-api-specification/blob/main/DEVELOPER_GUIDE.md), if applicable.
- [ ] Public documentation issue/PR [created](https://github.com/opensearch-project/documentation-website/issues/new/choose), if applicable.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

peternied commented:
@AndreKurait Thanks for the PR, looks like you need to run `./gradlew spotlessApply` and add a change log entry

AndreKurait commented:
> @AndreKurait Thanks for the PR, looks like you need to run `./gradlew spotlessApply` and add a change log entry

Updated spotless and added the changelog

AndreKurait commented:
@peternied, I've addressed all tests failures and added new tests to verify behavior. I've updated the behavior to allow updating creation_date during restore of an index with snapshot_restore too
